### PR TITLE
✨amp-inputmask initial commit

### PR DESCRIPTION
**Pull request by cvializ**
Implements #12079

`amp-inputmask` is implemented as a new extension used in conjunction with `amp-form`. `amp-form` installs the `amp-inputmask` service when `amp-form` attaches to the document and `amp-inputmask` has finished loading. 

`amp-inputmask` is created as a separate extension since not all documents that use `amp-form` will need input masking functionality, so they should not download the code for it.

This initial version supports custom masks using mask characters. Named masks are not yet implemented in this v0, but there are examples of their usage in `examples/masking.amp.html`.

/to @nainar @aghassemi for UI review
/to @honeybadgerdontcare for validator changes
/cc @calebcordry for form submit service usage and a small annotation change

### CODE REVIEW

File: extensions/amp-inputmask/validator-amp-inputmask.protoascii [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    # amp-inputmask has no associated tag which indicates usage of the extension.
+    # TODO(gregable): Implement a mechanism to associate attributes with
+    # extension usage and then set this to GRANDFATHERED or ERROR.
+    requires_usage: NONE
+  }
+  attr_lists: "common-extension-attrs"
+}
+tags: {
+  html_format: AMP
+  tag_name: "INPUT"
```
[the remark this answers is not in the document]
Comment (cvializ) [reply extensions/amp-inputmask/validator-amp-inputmask.protoascii]: Done, at line 30

File: extensions/amp-inputmask/0.1/amp-inputmask.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    this.masks_ = [];
+  }
+
+  /**
+   * Install the inputmask service and controllers.
+   */
+  install() {
+    const maskElements = this.ampdoc.getRootNode().querySelectorAll('[mask]');
+    iterateCursor(maskElements, element => {
+      user().assert(
```
Comment (nainar) [inline extensions/amp-inputmask/0.1/amp-inputmask.js]: Can't we check the assert just for the first element in the NodeList?

File: extensions/amp-form/0.1/form-submit-service.js [hunk 3]
```diff
@@ -19,7 +19,7 @@ import {Observable} from '../../../src/observable';
 
 /**
  * @typedef {{
- *   form: HTMLFormElement,
+ *   form: !HTMLFormElement,
```
Comment (nainar) [inline extensions/amp-form/0.1/form-submit-service.js]: why this change?

File: extensions/amp-inputmask/0.1/mask-impl.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    const win = element.ownerDocument.defaultView;
+
+    InputmaskDependencyLib = InputmaskDependencyLib ||
+        inputmaskDependencyFactory(win, doc);
+    Inputmask = Inputmask || inputmaskFactory(
+        InputmaskDependencyLib, win, doc, undefined);
+
+    Inputmask.extendDefaults({
+      supportsInputType: [
+        'text',
```
Comment (nainar) [inline extensions/amp-inputmask/0.1/mask-impl.js]: Add a comment what each of these input types means?

**Comment by aghassemi** [comment]
🎉  will review soon. meanwhile, please add a LICENSE and README.amp file for the third-party (something like this: https://github.com/ampproject/amphtml/tree/master/third_party/vega ) and open a ticket for security review.

File: extensions/amp-form/0.1/form-submit-service.js (hunk 3 above)
Comment (cvializ) [reply extensions/amp-form/0.1/form-submit-service.js]: The change is because the `formSubmitService.js#fire` function can only be triggered by a form that exists, so the type should be non-nullable. It also allows consumers to use the `FormSubmitEventDef#form` property without checking if it's non-null first.

Comment (calebcordry) [reply extensions/amp-form/0.1/form-submit-service.js]: Good call, thanks for tidying this up.

File: extensions/amp-inputmask/0.1/amp-inputmask.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+   */
+  install() {
+    const maskElements = this.ampdoc.getRootNode().querySelectorAll('[mask]');
+    if (maskElements.length) {
+      user().assert(
+          isExperimentOn(this.ampdoc.win, 'amp-inputmask'),
+          'Experiment amp-inputmask is disabled');
+    }
+
+    for (let i = 0; i < maskElements.length; i++) {
```
Comment (nainar) [inline extensions/amp-inputmask/0.1/amp-inputmask.js]: This can be pushed into the previous block and you can calculate & store `maskElements.length`

File: extensions/amp-form/0.1/amp-form.js [hunk 6]
```diff
@@ -355,6 +356,13 @@ export class AmpForm {
     });
   }
 
+  /** @private */
+  installInputMasking_() {
+    Services.inputmaskServiceForDoc(this.form_).then(inputmaskService => {
```
Comment (aghassemi) [inline extensions/amp-form/0.1/amp-form.js]: should this be `inputmaskServiceForDocOrNull` since extension is optional? I don't know what currently happens if inputmask extension is not included. My guess is promise is rejected and error gets logged as uncaught promise rejection?

File: extensions/amp-inputmask/0.1/amp-inputmask.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    /** @private {!Array<!TextMask>} */
+    this.masks_ = [];
+  }
+
+  /**
+   * Install the inputmask service and controllers.
+   */
+  install() {
+    const maskElements = this.ampdoc.getRootNode().querySelectorAll('[mask]');
```
Comment (aghassemi) [inline extensions/amp-inputmask/0.1/amp-inputmask.js]: we would need to redo this for when amp-list adds new inputs. So need something like
```
   doc.addEventListener(AmpEvents.DOM_UPDATE, () => {
      // query and install again
    });
```

File: extensions/amp-inputmask/0.1/amp-inputmask.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      const tm = new TextMask(maskElements[i]);
+      this.masks_.push(tm);
+    }
+  }
+
+  /**
+   * Remove the inpumask service and controllers.
+   */
+  uninstall() {
+    this.masks_.forEach(m => m.dispose());
```
Comment (aghassemi) [inline extensions/amp-inputmask/0.1/amp-inputmask.js]: per comment above, also needs to unlisten to DOM_UPDATE

File: extensions/amp-inputmask/0.1/mask-impl.js [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+      placeholder: '\u2000',
+      showMaskOnHover: false,
+      showMaskOnFocus: false,
+      noValuePatching: true,
+    };
+
+    if (NamedMasksToInputmask[mask]) {
+      config.alias = NamedMasksToInputmask[mask];
+    } else {
+      config.mask = () => mask
```
Comment (aghassemi) [inline extensions/amp-inputmask/0.1/mask-impl.js]: comment on what format this is generating would be nice, just a example would do IMO.

File: extensions/amp-inputmask/0.1/mask-impl.js [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+  /** @override */
+  dispose() {
+    this.controller_./*OK*/remove();
+    this.controller_ = null;
+  }
+}
+
+// TODO(cvializ): improve for unicode etc.
+const NONALPHANUMERIC_REGEXP = /[^A-Za-z0-9]/g;
```
Comment (aghassemi) [inline extensions/amp-inputmask/0.1/mask-impl.js]: would `^/w` work? I hope JS implementation considers unicode ! (it matches `_` too not sure if you want that or not)

File: extensions/amp-inputmask/0.1/text-mask.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+   * Add a hidden input to make the unmasked value available to the form
+   * @param {!HTMLFormElement} form
+   */
+  handleBeforeSubmit_(form) {
+    if (this.outputMode_ == OutputMode.ALPHANUMERIC) {
+      if (!this.hiddenInput_) {
+        const hidden = this.document_.createElement('input');
+        hidden.type = 'hidden';
+        const name = this.element_.name || this.element_.id;
+        hidden.name = `${name}-unmasked`; // TODO(cvializ): detect collisions
```
Comment (aghassemi) [inline extensions/amp-inputmask/0.1/text-mask.js]: is collision detection needed? IF author used the same name twice , the hidden should follow the same.

I think we should skip adding the hidden if there is no `name` on element.

File: extensions/amp-inputmask/README.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+  <!-- TODO(@cvializ) -->
+  <!-- <tr>
+    <td width="40%"><strong>Examples</strong></td>
+    <td>FILL THIS IN</td>
+  </tr> -->
+</table>
+
+## Behavior
+
+`amp-inputmask` enables the `mask` and `mask-output` attributes on `input` elements. These allow document authors to specify input masks for their form elements.
```
Comment (aghassemi) [inline extensions/amp-inputmask/README.md]: /cc @CrystalFaith for doc review.

**Comment by aghassemi** [comment]
Awesome work! just a few requests

File: extensions/amp-inputmask/0.1/mask-impl.js (hunk 10 above)
Comment (cvializ) [reply extensions/amp-inputmask/0.1/mask-impl.js]: Unfortunately it does not work. I ended up changing the RegExp here to a much more comprehensive one including the unicode character categories Letter and Number. http://www.unicode.org/reports/tr44/#General_Category_Values

File: extensions/amp-inputmask/0.1/text-mask.js (hunk 11 above)
Comment (cvializ) [reply extensions/amp-inputmask/0.1/text-mask.js]: I think collision detection is important so that we don't overwrite someone else's input if their form already has a `${name}-unmasked` input e.g.
```html
<form>
<input name="wow" mask="LLL-LLL" mask-output="alphanumeric">
<input name="wow-masked" type="email"><!-- would get shadowed -->
</form>
```
It's a low risk, but I think it's still something we should try to avoid

File: extensions/amp-inputmask/0.1/mask-impl.js (hunk 10 above)
Comment (aghassemi) [reply extensions/amp-inputmask/0.1/mask-impl.js]: great. sad that `/w` is Latin only :(

File: extensions/amp-inputmask/README.md (hunk 12 above)
Comment (CrystalFaith) [reply extensions/amp-inputmask/README.md]: The language looks really good! I see there is a todo - but its hard for me to sign off without a code sample. Its unclear to me how this is applied to an input(or replaces the input element?) - and then how I'm able to specify the mask attributes.

**Comment by cvializ** [comment]
@CrystalFaith thanks for the review! PTAL. I added some examples to clarify usage of the `mask` and `mask-output` attributes enabled by the `amp-inputmask` extension.
